client/src/api/block_builder/input_selection/new/mod.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+pub(crate) mod burn;
+pub(crate) mod helper;
+pub(crate) mod remainder;
+pub(crate) mod requirement;
+pub(crate) mod transition;
+
+use std::collections::HashSet;
+
+pub use burn::Burn;
+use helper::is_alias_state_transition;
+use remainder::remainder_output;
+pub use requirement::Requirement;
+use requirement::Requirements;
+
+use crate::{
+    block::{
+        address::{Address, AliasAddress, NftAddress},
+        output::{Output, OutputId},
+        protocol::ProtocolParameters,
+    },
+    error::{Error, Result},
+    secret::types::InputSigningData,
+};
+
+// TODO should ISA have its own error type? At least review errors.
+// TODO make methods actually take self? There was a mut issue.
+
+#[derive(Debug)]
+pub(crate) struct OutputInfo {
+    pub(crate) output: Output,
+    pub(crate) provided: bool,
+}
+
+/// Working state for the input selection algorithm.
+pub struct InputSelection {
+    available_inputs: Vec<InputSigningData>,
+    // TODO option needed ?
+    required_inputs: Option<HashSet<OutputId>>,
+    forbidden_inputs: HashSet<OutputId>,
+    selected_inputs: Vec<InputSigningData>,
+    outputs: Vec<OutputInfo>,
+    burn: Option<Burn>,
+    remainder_address: Option<Address>,
+    protocol_parameters: ProtocolParameters,
+    timestamp: u32,
+}
+
+impl InputSelection {
+    fn required_address(&self, input: &InputSigningData) -> Result<Option<Requirement>> {
+        // TODO burn?
+        // TODO unwrap or false?
+        // TODO this is only temporary to accommodate the current ISA.
+        let outputs = self
+            .outputs
+            .iter()
+            .map(|output| output.output.clone())
+            .collect::<Vec<_>>();
+        let is_alias_state_transition = is_alias_state_transition(input, &outputs)?.unwrap_or(false);
+        let (required_address, _) =
+            input
+                .output
+                .required_and_unlocked_address(self.timestamp, input.output_id(), is_alias_state_transition)?;
+
+        match required_address {
+            Address::Alias(alias_address) => Ok(Some(Requirement::Alias(*alias_address.alias_id()))),
+            Address::Nft(nft_address) => Ok(Some(Requirement::Nft(*nft_address.nft_id()))),
+            _ => Ok(None),
```

Comment:
Why are Ed25519 addresses not required?

Reply:
Well there are no further action like for alias and nft, what would you do ?

Reply:
Yes, just think that from reading this function alone you don't know why they aren't required
Maybe renaming it to required_alias_nft_addresses or at least adding a comment would help

Reply:
https://github.com/iotaledger/iota.rs/pull/1356/commits/89fb4b00d088317f9feec557685bf85adf08d6ad